dex/elrond_dex_pair/src/fee.rs

```diff
@@ -348,33 +342,66 @@ pub trait FeeModule:
         enabled: bool,
         fee_to_address: Address,
         fee_token: TokenIdentifier,
+        fee_percent: u64,
     ) -> SCResult<()> {
-        //require!(self.is_active(), "Not active");
         self.require_permissions()?;
-        let is_dest = self
-            .destination_map()
-            .keys()
-            .any(|dest_address| dest_address == fee_to_address);
+        require!(fee_percent != 0, "Fee percent cannot be zero");
+        require!(
+            fee_token.is_valid_esdt_identifier(),
+            "Not a valid esdt identifier"
+        );
+
+        let percent_sum: u64 = self.destination_map().iter().map(|x| x.1).sum();
+        let addr_token = (fee_to_address, fee_token);
+        let destination = self.destination_map().get(&addr_token);
 
         if enabled {
-            require!(!is_dest, "Is already a fee destination");
-            self.destination_map().insert(fee_to_address, fee_token);
+            require!(
+                percent_sum + fee_percent <= self.special_fee_percent().get(),
+                "Percent sum would exceed special fee percent"
+            );
+
+            match destination {
+                Some(old_percent) => {
+                    self.destination_map()
+                        .insert(addr_token, old_percent + fee_percent);
+                }
+                None => {
+                    self.destination_map().insert(addr_token, fee_percent);
+                }
+            }
         } else {
-            require!(is_dest, "Is not a fee destination");
-            let dest_fee_token = self.destination_map().get(&fee_to_address).unwrap();
-            require!(fee_token == dest_fee_token, "Destination fee token differs");
-            self.destination_map().remove(&fee_to_address);
+            require!(!destination.is_none(), "Destination does not exist");
+
+            let old_percent = destination.unwrap();
+            require!(
+                old_percent >= fee_percent,
+                "Old percent is less than given argument"
+            );
+
+            if old_percent > fee_percent {
+                self.destination_map()
+                    .insert(addr_token, old_percent - fee_percent);
+            } else {
+                self.destination_map().remove(&addr_token);
+            }
         }
         Ok(())
     }
 
     #[view(getFeeDestinations)]
-    fn get_fee_destinations(&self) -> MultiResultVec<(Address, TokenIdentifier)> {
+    fn get_fee_destinations(&self) -> MultiResultVec<(Address, TokenIdentifier, u64)> {
```

Comment:
This is easier to parse as `MultiResultVec<MultiResult3<Address, TokenIdentifier, u64>>`. The tuple will just concatenate all fields into one result.

Reply:
done